### PR TITLE
Code amelioration for card testing service

### PR DESCRIPTION
**Pull request by elazzabi**
This is an implementation of the second round of feedback for https://github.com/Automattic/woocommerce-payments/pull/4047 given the first PR was merged.

#### Code amelioration for fingerprinting service

This PR introduces some code amelioration for fingerprinting service

Questions for PR author:
- How can this code break? Code isn't used yet.
- What are we doing to make sure this code doesn't break? Fully covered with tests.

#### Testing instructions

* Changes are covered

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) : QA Testing Not Applicable

### CODE REVIEW

**Review by dmvrtx — APPROVED** [review]
Looks good to me! Thank you for looking into the suggestions!

**Comment by KarlisJ** [comment]
I can't get metadata to be added to the payment intent. 

I've manually set the feature to be anabled
```php
public function is_enabled(): bool {
		return true;
		return $this->wcpay_gateway->get_option( 'is_fraud_prevention_enabled' ) === 'yes';
	}
```
but it seems that execution of successful payment doesn't even trigger the method to be called. 

This
```php
die();
if ( Fraud_Prevention_Service::get_instance()->is_enabled() ) {
	$request['metadata']['fraud_prevention_data_available'] = true;
	$request['metadata']['fraud_prevention_data']           = Buyer_Fingerprinting_Service::get_instance()->get_hashed_data_for_order( $order_id );
}
```
is resulting in successful payment. 

What am I misunderstanding?

File: changelog/fix-fingerprinting-service [hunk 1]
```diff
@@ -0,0 +1,5 @@
+Significance: patch
+Type: fix
```
Comment (kalessil) [inline changelog/fix-fingerprinting-service:2]: Type: dev

**Review by kalessil — APPROVED** [review]
LGTM!

**Comment by KarlisJ** [comment]
We are still not doing anything from `create_and_confirm_intention` method. Is that intended?

**Comment by elazzabi** [comment]
@KarlisJ I think we should add the same metadata there too, but [I asked in Slack](p1649769240201959-slack-C02C00HCJ8J) in case other folks have a better idea of when `create_and_confirm_intention ` and `create_intention` are used...

File: includes/fraud-prevention/class-buyer-fingerprinting-service.php [hunk 2]
```diff
@@ -66,22 +51,18 @@ public static function set_instance( self $instance = null ) {
...
 	 * @return string[] An array of hashed data for an order.
 	 */
-	public function get_hashed_data_for_order( int $order_id ): array {
-		$order = wc_get_order( $order_id );
-
+	public function get_hashed_data_for_customer(): array {
 		return [
-			'shopper_ip_hash'        => self::hash_data_for_fraud_prevention( $order->get_customer_ip_address() ),
-			'shopper_useragent_hash' => self::hash_data_for_fraud_prevention( strtolower( $order->get_customer_user_agent() ) ),
+			'fraud_prevention_data_shopper_ip_hash' => $this->hash_data_for_fraud_prevention( WC_Geolocation::get_ip_address() ),
```
Comment (dmvrtx) [inline includes/fraud-prevention/class-buyer-fingerprinting-service.php:64]: In the future it'd be great for IP and User-Agent to be collected with JS from the browser client with the current implementation as a fallback.

Comment (dmvrtx) [reply includes/fraud-prevention/class-buyer-fingerprinting-service.php:64]: With this change singleton becomes useful for testing. E.g. you add two one-line methods to  `Buyer_Fingerprinting_Service`, i.e. `get_customer_ip` and `get_customer_useragent`. In the tests you create a mock-class and override these two methods with whatever values you need.

**Review by dmvrtx — APPROVED** [review]
Looks good to me. I left notes, but they would matter if we happen to extend this component.

File: includes/wc-payment-api/class-wc-payments-api-client.php [hunk 3]
```diff
@@ -241,6 +241,14 @@ public function create_and_confirm_intention(
 			$request['cvc_confirmation'] = $cvc_confirmation;
 		}
 
+		if ( Fraud_Prevention_Service::get_instance()->is_enabled() ) {
+			$request['metadata']['fraud_prevention_data_available'] = true;
+			$hashed_fingerprints                                    = Buyer_Fingerprinting_Service::get_instance()->get_hashed_data_for_customer();
+			foreach ( $hashed_fingerprints as $key => $value ) {
```
Comment (dpaun1985) [inline includes/wc-payment-api/class-wc-payments-api-client.php:247]: we can use `array_merge` here

File: includes/wc-payment-api/class-wc-payments-api-client.php [hunk 4]
```diff
@@ -274,7 +282,10 @@ public function create_intention(
 
 		if ( Fraud_Prevention_Service::get_instance()->is_enabled() ) {
 			$request['metadata']['fraud_prevention_data_available'] = true;
-			$request['metadata']['fraud_prevention_data']           = Buyer_Fingerprinting_Service::get_instance()->hash_data_for_fraud_prevention( $order_id );
+			$hashed_fingerprints                                    = Buyer_Fingerprinting_Service::get_instance()->get_hashed_data_for_customer();
+			foreach ( $hashed_fingerprints as $key => $value ) {
```
Comment (dpaun1985) [inline includes/wc-payment-api/class-wc-payments-api-client.php:286]: array_merge ?

File: tests/unit/fraud-prevention/test-class-buyer-fingerprinting-service.php [hunk 5]
```diff
@@ -41,14 +41,14 @@ public function test_it_hashes_using_sha512() {
 		$expected_hashed_string = hash( 'sha512', 'test_string' );
 
 		$this->assertSame( $result, $expected_hashed_string );
+		$this->assertTrue( is_string( $result ) );
```
Comment (dpaun1985) [inline tests/unit/fraud-prevention/test-class-buyer-fingerprinting-service.php:44]: this is already evaluated. `assertSame` check also the type

File: tests/unit/fraud-prevention/test-class-buyer-fingerprinting-service.php [hunk 6]
```diff
@@ -41,14 +41,14 @@ public function test_it_hashes_using_sha512() {
...
-		$order_hashes          = $this->buyer_fingerprinting_service->get_hashed_data_for_order( $order->get_order_number() );
+		$order_hashes          = $this->buyer_fingerprinting_service->get_hashed_data_for_customer();
 		$expected_hashed_array = [
-			'shopper_ip_hash'        => hash( 'sha512', $order->get_customer_ip_address() ),
-			'shopper_useragent_hash' => hash( 'sha512', $order->get_customer_user_agent() ),
+			'fraud_prevention_data_shopper_ip_hash' => hash( 'sha512', '127.0.0.1', false ),
+			'fraud_prevention_data_shopper_ua_hash' => hash( 'sha512', '', false ),
 		];
 
 		$this->assertEquals( $order_hashes, $expected_hashed_array );
```
Comment (dpaun1985) [inline tests/unit/fraud-prevention/test-class-buyer-fingerprinting-service.php:54]: better `assertSame` to check also the types

**Review by dpaun1985 — COMMENTED** [review]
Good job! Left a few non-important comments.

**Comment by elazzabi** [comment]
@KarlisJ @dpaun1985 I used a combination of both your feedback here https://github.com/Automattic/woocommerce-payments/pull/4120/commits/dad3659395b25be2f37fbd247d49e4dac4a17b5e

**Comment by elazzabi** [comment]
@dpaun1985 I agree, removed it in favor of a simple `array[] = `

File: includes/wc-payment-api/class-wc-payments-api-client.php [hunk 7]
```diff
@@ -277,7 +277,7 @@ public function create_intention(
 		$request['capture_method']       = $capture_method;
 
 		if ( Fraud_Prevention_Service::get_instance()->is_enabled() ) {
-			$request['metadata'] = array_merge( isset( $request['metadata'] ) ? $request['metadata'] : [], $this->get_fingerprint_metadata() );
+			$request['metadata'] = $this->get_fingerprint_metadata();
```
Comment (elazzabi) [inline includes/wc-payment-api/class-wc-payments-api-client.php:280]: According to `psalm` we aren't defining a `metadata` index before this line, so we directly assign `get_fingerprint_metadata` to the `metadata`

Comment (KarlisJ) [reply includes/wc-payment-api/class-wc-payments-api-client.php:280]: I think this is ok. If someone will make a change that introduces the metadata at some point then the tests should pick up and indicate the person to merge the metadatas.

**Comment by KarlisJ** [comment]
LGTM!

**Review by dpaun1985 — APPROVED** [review]
LGTM

**Pull request merged**